src/plugins/gds_mt/gds_mt_backend.cpp

```diff
@@ -0,0 +1,305 @@
+    // Full hardware concurrency often leads to context switching overhead
+    // and resource contention for storage I/O workloads
+    const size_t default_thread_count = std::max(1u, std::thread::hardware_concurrency() / 2);
+}
+
+nixlGdsMtEngine::nixlGdsMtEngine(const nixlBackendInitParams* init_params)
+    : nixlBackendEngine(init_params)
+{
+    gds_mt_utils_ = std::make_unique<gdsMtUtil>();
+
+    thread_count_ = default_thread_count;
+
+    // Read custom parameters if available
+    nixl_b_params_t* custom_params = init_params->customParams;
+    if (custom_params) {
+        // Configure thread_count
+        if (custom_params->count("thread_count") > 0) {
+            try {
+                size_t tcount = std::stoul((*custom_params)["thread_count"]);
+                if (tcount != 0) {
+                    thread_count_ = tcount;
+                }
+            } catch (const std::exception& e) {
+                throw std::runtime_error("GDS_MT: Invalid thread_count parameter: " + std::string(e.what()));
+            }
+        }
+    }
+    executor_ = std::make_unique<tf::Executor>(thread_count_);
+    NIXL_DEBUG << "GDS_MIT: thread count=" << thread_count_;
+}
+
+nixl_status_t nixlGdsMtEngine::registerMem(const nixlBlobDesc &mem,
+                                           const nixl_mem_t &nixl_mem,
+                                           nixlBackendMD* &out)
+{
+    auto md = std::make_unique<nixlGdsMtMetadata>();
+    md->type = nixl_mem;
+
+    switch (nixl_mem) {
+        case FILE_SEG: {
+            std::lock_guard<std::mutex> lock(mutex_);
+            auto it = gds_mt_file_map_.find(mem.devId);
+            if (it != gds_mt_file_map_.end()) {
+                // Reuse existing metadata and increment reference count
+                auto existing_md = it->second;
+                existing_md->ref_cnt++;
+                out = existing_md;
+                return NIXL_SUCCESS;
+            }
+
+            // Create and register new handle
+            try {
+                md->handle = std::make_shared<gdsMtFileHandle>(mem.devId);
+            } catch (const std::exception& e) {
+                NIXL_ERROR << "GDS_MT: failed to create file handle: " << e.what();
+                return NIXL_ERR_BACKEND;
+            }
+            md->ref_cnt = 1;
+            gds_mt_file_map_[mem.devId] = md.get();
+            break;
+        }
+
+        case VRAM_SEG: {
+            const cudaError_t error_id = cudaSetDevice(mem.devId);
+            if (error_id != cudaSuccess) {
+                NIXL_ERROR << "GDS_MT: error: cudaSetDevice returned "
+                           << cudaGetErrorString(error_id) << " for device ID " << mem.devId;
+                return NIXL_ERR_BACKEND;
+            }
+            [[fallthrough]];
+        }
+        case DRAM_SEG: {
+            try {
+                md->buf = std::make_unique<gdsMtMemBuf>((void *)mem.addr, mem.len, 0);
```

Comment:
Making metadata class stuct-like for simple cases is fine, but it quickly gets convoluted in more complicated cases like this one. How about fully initializing the class instance with a per-case constructor instead of creating the instance at the beggining on line 67 and filling it out in many different places?

I'm also wondering what is the best approach to simplify handling of handle and buf that only make sense for specific SEG types. Dedicated nixlBackendMD subclass per-type? Or maybe wrap them in std::variant? WDYT?